bee-message/src/parents.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! The parents module defines the core data type for storing the messages directly approved by a message.
+
+use crate::{MessageId, MessagePackError, MessageUnpackError, ValidationError, MESSAGE_ID_LENGTH};
+
+use bee_ord::is_unique_sorted;
+use bee_packable::{PackError, Packable, Packer, UnpackError, Unpacker};
+
+use bitvec::prelude::*;
+
+use core::ops::{Deref, RangeInclusive};
+
+use alloc::{vec, vec::Vec};
+
+/// The range representing the valid number of parents.
+pub const MESSAGE_PARENTS_RANGE: RangeInclusive<usize> = 1..=8;
+
+/// Minimum number of strong parents for a valid message.
+pub const MESSAGE_MIN_STRONG_PARENTS: usize = 1;
+
+/// An individual message parent, which can be categorized as "strong" or "weak".
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(tag = "type", content = "data")
+)]
+pub enum Parent {
```

Comment:
Wondering if it should derive/impl Packable itself instead of implicitly doing it in Parents

Reply:
I'm not sure we can do this: the type is determined by the preceding `u8` bitfield that uses 0/1 to indicate `Weak`/`Strong` parents following. This is then followed by each corresponding `MessageId`. So, if unpacking each `Parent` individually, there is no way to know what variant it is unless you know the bitfield (which we can't provide to the `unpack` method).